src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -156707,5 +156707,56 @@ public static void DoWork(List<Goo> list, string unrelatedObject)
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x.Bar").WithLocation(57, 36)
                 );
         }
+
+        [Fact, WorkItem(65050, "https://github.com/dotnet/roslyn/issues/65050")]
+        public void UnreachableWhenClause_SwitchExpression()
+        {
+            var source = """
+                #nullable enable
+                class C
+                {
+                    string? M(bool b, string s)
+                    {
+                        return b switch
+                        {
+                            _ when s.Length == 0 => s,
+                            _ => null,
+                            true when s.Length > 0 => s,
+                        };
+                    }
+                }
+                """;
+            CreateCompilation(source).VerifyDiagnostics(
```

Comment:
It might be good to check the nullability info coming from public API as well, a la NullablePublicAPITests. Probably, everything will always show as non-null due to being visited in an unreachable state. But still may be useful for completeness.